File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/election/ElectionResultQuestion.java
package com.github.dedis.popstellar.model.network.method.message.data.election;

import androidx.annotation.NonNull;

import com.github.dedis.popstellar.model.Immutable;

import java.util.*;

@Immutable
public class ElectionResultQuestion {

  private final String id;
  private final Set<QuestionResult> result;

  public ElectionResultQuestion(@NonNull String id, @NonNull Set<QuestionResult> result) {
    if (result.isEmpty()) {
      throw new IllegalArgumentException();
    }
    this.id = id;
    this.result = new HashSet<>(result);
  }

  public @NonNull String getId() {
    return id;
  }

  public @NonNull Set<QuestionResult> getResult() {
    return result;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    ElectionResultQuestion that = (ElectionResultQuestion) o;

    return Objects.equals(id, that.id) && Objects.equals(result, that.result);
  }

  @Override
  public int hashCode() {
    return Objects.hash(id, result);
  }

  @NonNull
  @Override
  public String toString() {
    return "ElectionResultQuestion{"
        + "id='"
        + id
        + '\''
        + ", result="
        + Arrays.toString(result.toArray())
        + '}';
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/election/ElectionResultQuestion.kt
package com.github.dedis.popstellar.model.network.method.message.data.election

import com.github.dedis.popstellar.model.Immutable
import com.github.dedis.popstellar.utility.MessageValidator.verify
import java.util.Objects

@Immutable
class ElectionResultQuestion(id: String, result: Set<QuestionResult>) {
  val id: String
  val result: Set<QuestionResult>

  init {
    // Set can't have duplicates, so no need to check for duplicates
    // QuestionResult are already validated when constructed
    verify().stringNotEmpty(id, "election id").listNotEmpty(result.toList())

    this.id = id
    this.result = HashSet(result)
  }

  override fun equals(other: Any?): Boolean {
    if (this === other) {
      return true
    }
    if (other == null || javaClass != other.javaClass) {
      return false
    }
    val that = other as ElectionResultQuestion
    return id == that.id && result == that.result
  }

  override fun hashCode(): Int {
    return Objects.hash(id, result)
  }

  override fun toString(): String {
    return "ElectionResultQuestion{id='$id', result=${result.toTypedArray().contentToString()}}"
  }
}
